src/common/utils/pipeline.ts

```diff
@@ -1,17 +1,62 @@
-import { AxiosInstance } from 'axios';
+import AppService from '../../app-config/service';
+import { StringTemplateBuilder, TemplateValues } from './template-builder';
+
+class PipelineAbortedError extends Error {
+  constructor(deployment_id: string, deployment_link: string) {
+    super();
+
+    this.message = `Deployment ${deployment_id} was aborted\n See deployment log for more details: ${deployment_link}`;
+  }
+}
+
+class DeploymentFailedError extends Error {
+  constructor(pipeline_id: string, deployment_links: string[]) {
+    super();
+
+    const deployment_string = deployment_links.length > 1
+      ? `${deployment_links.length} deployments`
+      : '1 deployment';
+    const listified_link_string = deployment_links.map((s: string) => `- ${s}`).join('\n');
+    this.message = `Pipeline ${pipeline_id} failed because ${deployment_string} failed:\n${listified_link_string}`;
+  }
+}
+
+const deploymentUrlBuilder = new StringTemplateBuilder(
+  ['app_host', 'account', 'environment', 'deployment'],
+  (p: TemplateValues): string => `${p.app_host}/${p.account}/environments/${p.environment}/deployments/${p.deployment}`
+);
 
 export class PipelineUtils {
 
   static POLL_INTERVAL = 10000;
 
-  static async pollPipeline(api: AxiosInstance, pipeline_id: string): Promise<any> {
+  static async pollPipeline(app: AppService, pipeline_id: string): Promise<any> {
     return new Promise((resolve, reject) => {
       let poll_count = 0;
       const poll = setInterval(async () => {
-        const { data: pipeline } = await api.get(`/pipelines/${pipeline_id}`);
+        const { data: pipeline } = await app.api.get(`/pipelines/${pipeline_id}`);
         if (pipeline.failed_at || poll_count > 180) {  // Stop checking after 30min (180 * 10s)
+          const deploymentUrl = deploymentUrlBuilder.with({
+            app_host: app.config.app_host,
+            account: pipeline.environment.account.name,
+            environment: pipeline.environment.name,
+          });
           clearInterval(poll);
-          reject(new Error('Pipeline failed'));
+          // Query deployments for pipline to determine cause of failure
+          const { data: deployments } = await app.api.get(`/pipelines/${pipeline.id}/deployments`);
+
+          // Check if the deployment failed due to a user aborting the deployment
+          const aborted_deployments = deployments.filter((d: any) => d.aborted_at);
```

Comment:
This check would need to be that every deployment is either applied_at or aborted_at. Since if a deployment hard fails the rest of the deployments that depend on it get aborted.

```ts
// common/utils/deployment.ts
const { data: deployments } = await app.api.get(`/pipelines/${pipeline.id}/deployments`) as Deployment[];
const pipeline_aborted = deployments.every(d => d.applied_at || d.aborted_at);
```

Reply:
I don't think that's quite right- if every deployment was successful and none were aborted, `pipeline_aborted` would still evaluate to `true` in your example. If a deployment fails and the following deployments are all aborted, then should it be find to just filter by `aborted_at` and check if the list is empty or not

Reply:
It'd only work in this block since the pipeline.failed_at is set